### PR TITLE
uid registry

### LINKED ISSUE #839: Implicit conversions for `FlagCollection`
cpplint enforces that single-argument constructors be marked explicit. This doesn't make sense in a heavily templated code because, for example, what is the meaning of marking this explicit?
```C++
template<typename T>
MyClass(const T &arg)
```
worse, marking it `explicit` breaks template resolution, as I discovered in #831 when trying to implement a simpler constructor based on @lroberts36 's simple concepts machinery. I suggest we remove this check from our copy of cpplint. (However, I wanted to make an issue, rather than just do it, as I think this one merits collaboration consensus.)

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

This PR builds on and depends on #831 . Here I take the unique identifier machinery for variables and thread it through packing so that key logic now uses lists of unique IDs, which are integers, rather than strings. This is essentially a performance boost for now, though I'm not sure how much it matters in practice. It probably helps in the limit of many small meshblocks. However, in the long term, it exposes these unique identifiers for use in sparse packs.

@lroberts36 also expressed interest in reorganizing the structure of this PR slightly to further decouple the unique IDs from `MeshBlockData`, so this is a place for us to collaborate on that.

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] Docs build
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Review by lroberts36 — APPROVED** [review]
This LGTM @Yurlungur. Please take a look and see if you find any issues with my changes. 

I think I misunderstood from our conversation where the unique ID reqistry was being stored. Having it as a static variable in `CellVariable` seems like a good solution that is nicely decoupled. I was thinking that this was somehow stored in `MeshBlockData` objects, but that is clearly not the case. I am not sure the unique IDs are consistent across ranks, but I guess that doesn't matter since they aren't shared across ranks. 

The only real change I ended up making was exposing a static method in `CellVariable` by which you can add variable names to the name to id map externally and calling that from within `AddFieldImpl` in `StateDescriptor` so variable unique IDs are available as early as possible and can be queried without access to a cell variable itself. This should be all I need for the sparse pack update I think. 

I also fixed #839 here, as well as removed the compiler warnings about __host__ __device__ destructors by explicitly declaring the destructors. I also slightly renamed a couple of things. 

I will thread this stuff into the `SparsePack`s in a separate PR.

File: src/interface/metadata.hpp
Comment (Yurlungur) [inline src/interface/metadata.hpp:222]: Ah mystery solved!

File: src/interface/metadata.hpp
Comment (Yurlungur) [inline src/interface/metadata.hpp:226]: Thanks for the catch @lroberts36

File: src/interface/variable.hpp
Comment (Yurlungur) [inline src/interface/variable.hpp:123]: It's a little unintuitive that this *sets* the unique ID if it's not present (my fault for the naming) but I think that's probably fine. It's a greedy get.

File: src/interface/variable.hpp
Comment (Yurlungur) [inline src/interface/variable.hpp]:
```suggestion
  // evaluate than a string. Safe so long as the number of MPI ranks
  // does not change while the code is running (restarts are fine).
```

**Review by Yurlungur — COMMENTED** [review]
Thanks @lroberts36 LGTM. I removed the `[WIP]` and will add changelog and check formatting one last time. Some comments below.

**Comment by Yurlungur** [comment]
@pgrete this is ready for review and it needs one more pair of eyes. Do you think Athena++ eyes are necessary? 

Alternatively does anyone else want to give this a quick look? @bprather @pdmullen @forrestglines ?

**Comment by Yurlungur** [comment]
Ok it looks like sparse advection regression is failing. I'll have a look at this to see what's going on.

File: src/interface/meshblock_data.hpp
Comment (brryan) [inline src/interface/meshblock_data.hpp:149]: Maybe a `PARTHENON_DEBUG_THROWS(varUidMap_.count(uid), "Variable ID not found!");`?

File: src/interface/meshblock_data.hpp
Comment (brryan) [inline src/interface/meshblock_data.hpp]: Is it worth promoting this magic number to a named quantity e.g. `vpack_types::invalidID`? I wouldn't necessarily be able to guess whether the right magic number is `0` or `-1` since I have to look elsewhere to know that the unique ID is unsigned.

File: src/utils/unique_id.hpp
Comment (brryan) [inline src/utils/unique_id.hpp:22]: I do like that this is promoted to a new name

**Review by brryan — APPROVED** [review]
LGTM -- non-blocking comments only

File: src/interface/meshblock_data.hpp
Comment (Yurlungur) [reply src/interface/meshblock_data.hpp:149]: Just realized this isn't necessary because `map.at(val)` has bounds checking. I added it anyway though for a nicer error message.

File: src/interface/meshblock_data.hpp
Comment (Yurlungur) [reply src/interface/meshblock_data.hpp]: Done

File: src/CMakeLists.txt
Comment (Yurlungur) [inline src/CMakeLists.txt]: This change was introduced to support backwards compatibility for `std::format`, which was introduced in #848 . Neither @pgrete nor I noticed that 

> Using this library may require additional compiler/linker options. GNU implementation prior to 9.1 requires linking with -lstdc++fs and LLVM implementation prior to LLVM 9.0 requires linking with -lc++fs. 

I take responsibility. Wish I'd noticed the fine print.


This set of cmake generator expressions should enable older versions of gcc/clang to link appropriately. Probably should be tested more carefully but it seems to work for us locally.

**Comment by Yurlungur** [comment]
> Ok it looks like sparse advection regression is failing. I'll have a look at this to see what's going on.

Thanks to @lroberts36 I believe this is resolved. At this point it would be good to get review from someone on the AthenaPK or Kharma sides. @pgrete @BenWibking @forrestglines @bprather

File: src/bvals/cc/build_boundary_buffers.cpp
Comment (forrestglines) [inline src/bvals/cc/build_boundary_buffers.cpp:50]: This parameter doesn't get used? Any reason to keep this around or does this function fulfill an interface requirement?

File: src/CMakeLists.txt
Comment (forrestglines) [reply src/CMakeLists.txt]: Is testing older GCC's in the CI worth it to catch these sorts of issues? Probably not?

**Comment by Yurlungur** [comment]
Ping @pgrete and @forrestglines this PR has been waiting for merge for over a month. Since it has two approving PRs from LANL folks, I am going to enable auto-merge if you don't say something by the end of the day next Monday.

**Comment by forrestglines** [comment]
> Ping @pgrete and @forrestglines this PR has been waiting for merge for over a month. Since it has two approving PRs from LANL folks, I am going to enable auto-merge if you don't say something by the end of the day next Monday.
I'm halfway through a review on this one, we discussed this in the meeting today. I should have a review done today

**Comment by Yurlungur** [comment]
Cool. Thanks!

File: src/CMakeLists.txt
Comment (c-prather) [reply src/CMakeLists.txt]: This doesn't compile under NVHPC, and probably won't under many compiler IDs we otherwise support (IntelLLVM != Clang, AOCC != Clang, IBM, etc, see [list](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_ID.html)).

Would it be possible to just try to link both libraries, and ignore failures?

**Review by c-prather — APPROVED** [review]
Using this in KHARMA now and it passes the tests.  Should be marked as breaks-downstream since it does change some return values I'd consider part of the "interface," which I was using.
Also, it looks like explicitly constructed `std::vector<MetadataFlag>()` are no longer valid in `GetByFlag`, which isn't a problem (use the bare initializer list) but is an incompatibility.

No other comments except the one about the `std::filesystem` bugfix

File: src/CMakeLists.txt
Comment (c-prather) [reply src/CMakeLists.txt]: After messing with `find_library` a while, I've decided that libraries are hard even if you don't special-case compilers, and it would be much easier to use something like `find_package(Filesystem)`, i.e., which someone else has already written.

Luckily these exist, I've had good luck with [this one](https://github.com/lethal-guitar/RigelEngine/blob/master/cmake/Modules/FindFilesystem.cmake), released under BSD license.  (There is a rather hiliarously old CMake [bug](https://gitlab.kitware.com/cmake/cmake/-/issues/17834) to make something like this standard). Happy to add the file and the couple of lines required to switch to this, if I get the okay.

**Comment by Yurlungur** [comment]
> Also, it looks like explicitly constructed std::vector<MetadataFlag>() are no longer valid in GetByFlag, which isn't a problem (use the bare initializer list) but is an incompatibility.

Interesting. In my original design this worked. I assume it no longer works because we agreed to forbid implicit type conversions in our constructors. It should be possible to add overloads if this is desired.

File: src/CMakeLists.txt
Comment (Yurlungur) [reply src/CMakeLists.txt]: :+1: I'm in favor of this, and thanks for finding it! Please feel free to push directly to this branch.

File: src/interface/variable.hpp
Comment (forrestglines) [reply src/interface/variable.hpp:123]: Is there a conventional name for a greedy get? It seems like a word that should exist but I couldn't think of one.

**Review by forrestglines — APPROVED** [review]
Looks good to me. This shouldn't interfere with the current methods we use in AthenaPK using strings to get variables, correct?

UIDs should be the same from run to run so it won't interfere with debugging, right?

**Comment by Yurlungur** [comment]
> This shouldn't interfere with the current methods we use in AthenaPK using strings to get variables, correct?
> 
> UIDs should be the same from run to run so it won't interfere with debugging, right?

Yep and yep. :)

File: src/bvals/cc/build_boundary_buffers.cpp
Comment (Yurlungur) [reply src/bvals/cc/build_boundary_buffers.cpp:50]: It fulfills an interface requirement.

File: src/interface/variable.hpp
Comment (Yurlungur) [reply src/interface/variable.hpp:123]: Yeah same.

File: src/CMakeLists.txt
Comment (Yurlungur) [reply src/CMakeLists.txt]: > Is testing older GCC's in the CI worth it to catch these sorts of issues? Probably not?

It might be worth it... but the problem will just go away with time.

**Comment by Yurlungur** [comment]
Thanks @forrestglines ! And thanks @bprather for the cmake fix. I'll pull the trigger.

**Pull request merged**
